docs/adr/0002-release-and-branching-strategy.md

```diff
@@ -0,0 +1,208 @@
+* Provides clear structure for developer workflow for issues where development dependencies are identified.
+* Does not preclude future changes to release process.
+* Does not preclude the ability to release a breaking change "out of order" (akin to possibilities in Option 3), should it be necessary.
+
+### Negative Consequences
+
+* Some effort required to document branching strategy.
+* Requires additional onboarding time to become familiar branching strategy.
+* Requires process and tooling changes around PRs and the CI pipeline (see Pros and Cons of [Modified Gitflow](#modified-gitflow)).
+
+## Pros and Cons of the Options
+
+### Option 1: Current process: Formulate releases ad-hoc
+* Good, because there is little process overhead.
+* Good, because the release process is low commitment. Releases can be performed at any time based on discretion and convenience.
+* Bad, because moderate maintainer overhead to keep track of development dependencies is required.
+* Bad, because it does not provide clear focus/priority for backlog.
+
+#### Specific Pros and Cons in combination with compatible branching strategies
+* A. [Default](#default)
+  * Good, because release process & branching strategy do not change.
+  * Bad, because it does not address the [problem statement](#context-and-problem-statement) by not providing a clear path for navigating development dependencies.
+* B. [Standard Gitflow](#standard-gitflow)
+  * Bad, because it adds process overhead while not addressing the [problem statement](#context-and-problem-statement) by not providing a clear path for navigating development dependencies.
+* C. [Modified Gitflow](#modified-gitflow)
+  * Good, because it scales with contributors and development dependencies, directly addressing the problem statement. 
+
+### Option 2: Release all changes immediately
+* Good, because it lifts restrictions on the developer workflow that breaking or blocking changes would ordinarily cause.
+* Good, because developer overhead is significantly reduced by removing decision energy around planning work.
+* Good, because maintainer overhead is reduced by removing decision energy around planning releases.
+* Good, because it facilitates a consistent release process. There are no conditional steps in the process, just release and increment the appropriate version accordingly.
+* Good, because it eliminates the need for prioritizing the backlog.
+* Bad, because the stability of the library for consumers becomes a secondary concern.
+* Bad, because more frequent breaking releases.
+* Bad, because being a consumer of `react-uswds` may feel chaotic to endure breaking releases between features.
+* Bad, because the level of support in relation to USWDS becomes unclear as features may be released out of "chronological" order (i.e. `react-uswds` depends on 2.10.0 for a released 2.10.0 component before all 2.9.0 components are completed).
```

Comment:
Do we have any more docs on this? It seems unfortunate to me that if one of our projects needs only one component from a future release we can't release a version of react-uswds with it until we've implemented all components from all prior releases. Or do I misunderstand?

Reply:
Said another way, IMHO our compatibility promise should be that *for all components we have implemented* they will be compatible with a given USWDS release, not that we have implemented all components from a particular USWDS release.

Reply:
I don't think we formally have more docs on this. Most of these are coming from discussions I've been a part of. That's an interesting way to look at it though! I don't have reservations with the idea of "sniping" USWDS components by updating our version to do so, so long as we're still intentional about picking up with smaller changes and fixes that often accompany USWDS releases. We've been more diligent about that lately, so that practice would definitely help.

In that scenario I can also see USWDS changes for a particular release being specific to an un-implemented React USWDS component, which we would also just have to be diligent about making issues in the backlog for, or updating the "new component" issue. That way we don't miss out on things to maintain parity while "skipping" components.

Framing the "compatibility promise" a certain way and documenting that publicly would be really helpful here. Now I'm not exactly sure if we've implicitly adopted one, and whether its even the right one to use.

Reply:
I think that compatibility promise has a big impact on the branching strategy and why it is or is not necessary. I'm not trying to derail this ADR, but I am trying to make sure we document the assumptions that go into the ADR. I wonder if it's possible compatibility promise is another pre-requisite ADR?

Reply:
No worries! Fully understood and you're absolutely right! I _do_ think the compatibility promise makes or breaks some of these options, and it shouldn't be left as an assumption as I had done. I'm leaning in the direction of a pre-requisite ADR. On its own, I don't think it would be very large.